File: crates/warp-cli/src/ai_provider.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{mpsc, RwLock};
use tokio::time::{sleep, Duration};

#[derive(Debug, Clone, PartialEq)]
pub enum MessageRole {
    System,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ContentPartType {
    Text,
    Image,
    ToolCall,
}

#[derive(Debug, Clone)]
pub struct ContentPart {
    pub part_type: ContentPartType,
    pub text: String,
    pub call_id: String,
    pub name: String,
    pub args: String,
}

#[derive(Debug, Clone)]
pub struct Message {
    pub role: MessageRole,
    pub content: Vec<ContentPart>,
}

#[derive(Debug, Clone)]
pub struct Context {
    pub system_prompt: String,
    pub messages: Vec<Message>,
    pub tools: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct Model {
    pub provider_id: String,
    pub model_id: String,
    pub context_window: usize,
    pub max_tokens: usize,
    pub supports_images: bool,
    pub supports_tools: bool,
}

#[derive(Debug, Clone)]
pub enum EventType {
    Start,
    TextDelta,
    ToolCallStart,
    ToolCallDelta,
    Done,
    Error,
}

#[derive(Debug, Clone)]
pub struct AssistantMessageEvent {
    pub event_type: EventType,
    pub data: String,
    pub error: Option<String>,
}

#[async_trait::async_trait]
pub trait ApiProvider: Send + Sync {
    async fn stream(&self, model: Model, ctx: Context) -> mpsc::Receiver<AssistantMessageEvent>;
}

pub struct DummyProvider {
    pub name: String,
}

#[async_trait::async_trait]
impl ApiProvider for DummyProvider {
    async fn stream(&self, model: Model, ctx: Context) -> mpsc::Receiver<AssistantMessageEvent> {
        let (tx, rx) = mpsc::channel(100);
        let name = self.name.clone();

        tokio::spawn(async move {
            let _ = tx.send(AssistantMessageEvent { event_type: EventType::Start, data: "".to_string(), error: None }).await;

            let phrase = format!("This is a simulated streaming response from {} using model {}. ", name, model.model_id);
            let words: Vec<&str> = phrase.split_whitespace().collect();

            for word in words {
                let _ = tx.send(AssistantMessageEvent {
                    event_type: EventType::TextDelta,
                    data: format!("{} ", word),
                    error: None,
                }).await;
                sleep(Duration::from_millis(10)).await;
            }

            if !ctx.tools.is_empty() {
                let _ = tx.send(AssistantMessageEvent { event_type: EventType::ToolCallStart, data: "mcp_shell".to_string(), error: None }).await;
                let _ = tx.send(AssistantMessageEvent { event_type: EventType::ToolCallDelta, data: "{\"cmd\": \"echo hello\"}".to_string(), error: None }).await;
            }

            let _ = tx.send(AssistantMessageEvent { event_type: EventType::Done, data: "".to_string(), error: None }).await;
        });

        rx
    }
}

pub struct ApiProviderRegistry {
    providers: Arc<RwLock<HashMap<String, Arc<dyn ApiProvider>>>>,
    models: Arc<RwLock<HashMap<String, Model>>>,
}

impl ApiProviderRegistry {
    pub fn new() -> Self {
        Self {
            providers: Arc::new(RwLock::new(HashMap::new())),
            models: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn register_provider(&self, id: String, provider: Arc<dyn ApiProvider>) {
        let mut p = self.providers.write().await;
        p.insert(id.clone(), provider);
        println!("[ApiRegistry] Registered provider: {}", id);
    }

    pub async fn register_model(&self, model: Model) {
        let key = format!("{}:{}", model.provider_id, model.model_id);
        let mut m = self.models.write().await;
        m.insert(key.clone(), model);
        println!("[ApiRegistry] Registered model: {}", key);
    }

    pub async fn execute_stream(&self, provider_id: &str, model_id: &str, ctx: Context) -> Result<mpsc::Receiver<AssistantMessageEvent>, String> {
        let key = format!("{}:{}", provider_id, model_id);

        let models = self.models.read().await;
        let model = models.get(&key).cloned().ok_or_else(|| format!("Model {} not found", key))?;
        drop(models);

        let providers = self.providers.read().await;
        let provider = providers.get(provider_id).cloned().ok_or_else(|| format!("Provider {} not found", provider_id))?;
        drop(providers);

        Ok(provider.stream(model, ctx).await)
    }
}

File: warp-java/src/main/java/dev/warp/AiProvider.java
package dev.warp;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Flow;
import java.util.concurrent.SubmissionPublisher;
import java.util.concurrent.CompletableFuture;

public class AiProvider {

    public enum MessageRole { SYSTEM, USER, ASSISTANT, TOOL }

    public enum ContentPartType { TEXT, IMAGE, TOOL_CALL }

    public static class ContentPart {
        public ContentPartType type;
        public String text = "";
        public String callID = "";
        public String name = "";
        public String args = "";
    }

    public static class Message {
        public MessageRole role;
        public List<ContentPart> content = new ArrayList<>();
    }

    public static class Context {
        public String systemPrompt = "";
        public List<Message> messages = new ArrayList<>();
        public List<String> tools = new ArrayList<>();
    }

    public static class Model {
        public String providerID = "";
        public String modelID = "";
        public int contextWindow;
        public int maxTokens;
        public boolean supportsImages;
        public boolean supportsTools;
    }

    public enum EventType { START, TEXT_DELTA, TOOL_CALL_START, TOOL_CALL_DELTA, DONE, ERROR }

    public static class AssistantMessageEvent {
        public EventType type;
        public String data = "";
        public Exception error;

        public AssistantMessageEvent(EventType type) { this.type = type; }
        public AssistantMessageEvent(EventType type, String data) { this.type = type; this.data = data; }
    }

    public interface ApiProvider {
        Flow.Publisher<AssistantMessageEvent> stream(Model model, Context ctx);
    }

    public static class DummyProvider implements ApiProvider {
        public String name = "Dummy";

        @Override
        public Flow.Publisher<AssistantMessageEvent> stream(Model model, Context ctx) {
            SubmissionPublisher<AssistantMessageEvent> publisher = new SubmissionPublisher<>();

            CompletableFuture.runAsync(() -> {
                publisher.submit(new AssistantMessageEvent(EventType.START));

                String[] words = ("This is a simulated streaming response from " + name + " using model " + model.modelID + ". ").split(" ");
                for (String w : words) {
                    publisher.submit(new AssistantMessageEvent(EventType.TEXT_DELTA, w + " "));
                    try { Thread.sleep(10); } catch (InterruptedException ignored) {}
                }

                if (!ctx.tools.isEmpty()) {
                    publisher.submit(new AssistantMessageEvent(EventType.TOOL_CALL_START, "mcp_shell"));
                    publisher.submit(new AssistantMessageEvent(EventType.TOOL_CALL_DELTA, "{\"cmd\": \"echo hello\"}"));
                }

                publisher.submit(new AssistantMessageEvent(EventType.DONE));
                publisher.close();
            });

            return publisher;
        }
    }

    public static class ApiProviderRegistry {
        private final Map<String, ApiProvider> providers = new ConcurrentHashMap<>();
        private final Map<String, Model> models = new ConcurrentHashMap<>();

        public void registerProvider(String id, ApiProvider provider) {
            providers.put(id, provider);
            System.out.println("[ApiRegistry] Registered provider: " + id);
        }

        public void registerModel(Model model) {
            String key = model.providerID + ":" + model.modelID;
            models.put(key, model);
            System.out.println("[ApiRegistry] Registered model: " + key);
        }

        public Flow.Publisher<AssistantMessageEvent> executeStream(String providerID, String modelID, Context ctx) throws Exception {
            String key = providerID + ":" + modelID;
            Model model = models.get(key);
            if (model == null) {
                throw new Exception("Model " + key + " not found");
            }

            ApiProvider provider = providers.get(providerID);
            if (provider == null) {
                throw new Exception("Provider " + providerID + " not found");
            }

            return provider.stream(model, ctx);
        }
    }
}
